pkg/storage/utils/decomposedfs/decomposedfs.go

```diff
@@ -465,6 +492,159 @@ func (fs *Decomposedfs) Download(ctx context.Context, ref *provider.Reference) (
+	// TODO check filters
+
+	// TODO when a space symlink is broken delete the space for cleanup
+	// read permissions are deduced from the node?
+
+	// TODO for absolute references this actually requires us to move all user homes into a subfolder of /nodes/root,
+	// e.g. /nodes/root/<space type> otherwise storage space names might collide even though they are of different types
+	// /nodes/root/personal/foo and /nodes/root/shares/foo might be two very different spaces, a /nodes/root/foo is not expressive enough
+	// we would not need /nodes/root if access always happened via spaceid+relative path
+
+	var (
+		spaceType = "*"
+		spaceID   = "*"
+	)
+
+	for i := range filter {
+		switch filter[i].Type {
+		case provider.ListStorageSpacesRequest_Filter_TYPE_SPACE_TYPE:
+			spaceType = filter[i].GetSpaceType()
+		case provider.ListStorageSpacesRequest_Filter_TYPE_ID:
+			parts := strings.SplitN(filter[i].GetId().OpaqueId, "!", 2)
+			if len(parts) == 2 {
+				spaceID = parts[1]
+			}
+		}
+	}
+
+	// build the glob path, eg.
+	// /path/to/root/spaces/personal/nodeid
+	// /path/to/root/spaces/shared/nodeid
+	matches, err := filepath.Glob(filepath.Join(fs.o.Root, "spaces", spaceType, spaceID))
+	if err != nil {
+		return nil, err
+	}
+
+	spaces := make([]*provider.StorageSpace, 0, len(matches))
+
+	u, ok := user.ContextGetUser(ctx)
+	if !ok {
+		appctx.GetLogger(ctx).Debug().Msg("expected user in context")
+		return spaces, nil
+	}
+
+	for i := range matches {
+		// always read link in case storage space id != node id
+		if target, err := os.Readlink(matches[i]); err != nil {
+			appctx.GetLogger(ctx).Error().Err(err).Str("match", matches[i]).Msg("could not read link, skipping")
+			continue
+		} else {
+			n, err := node.ReadNode(ctx, fs.lu, filepath.Base(target))
+			if err != nil {
+				appctx.GetLogger(ctx).Error().Err(err).Str("id", filepath.Base(target)).Msg("could not read node, skipping")
+				continue
+			}
+			owner, err := n.Owner()
+			if err != nil {
+				appctx.GetLogger(ctx).Error().Err(err).Interface("node", n).Msg("could not read owner, skipping")
+				continue
+			}
+
+			// TODO apply more filters
+
+			space := &provider.StorageSpace{
+				// FIXME the driver should know its id move setting the spaceid from the storage provider to the drivers
+				//Id: &provider.StorageSpaceId{OpaqueId: "1284d238-aa92-42ce-bdc4-0b0000009157!" + n.ID},
+				Root: &provider.ResourceId{
+					// FIXME the driver should know its id move setting the spaceid from the storage provider to the drivers
+					//StorageId: "1284d238-aa92-42ce-bdc4-0b0000009157",
+					OpaqueId: n.ID,
+				},
+				Name:      n.Name,
+				SpaceType: filepath.Base(filepath.Dir(matches[i])),
+				// Mtime is set either as node.tmtime or as fi.mtime below
+			}
+
+			if space.SpaceType == "share" {
+				if utils.UserEqual(u.Id, owner) {
+					// do not list shares as spaces for the owner
+					continue
+				}
+				space.Name = n.Name
```

Comment:
`space.Name = n.Name` this will already be set previously

Reply:
good catch!